frontend/src/components/TaskCardModal.js

```diff
@@ -126,6 +129,22 @@ const TaskCardModal = (props) => {
         dispatch(addAlert({ severity: 'error', message: 'Something went wrong!' }))
       })
   }
+  const unassignUser = (user) => {
+    const payload = {
+      userId: user.id,
+    }
+    apiClient.post(`/api/tasks/${props.taskId}/unassign_user`, payload)
+      .then((response) => {
+        // successful request
+        mutateTask({ ...task, users: [...task?.users || [], user] })
```

Comment:
To nie ma sensu. Ten kod, który wywołuje odpięcie usera od zadania na chwilę zanim otrzyma odpowiedź z backendu z nowymi danymi taska dodaje drugi raz usera który jest odpinany do listy przypisanych.

Efekt będzie taki, że przy odpinaniu usera z wolnym internetem (lub z obciążonym serwerem) na chwilę pojawi się dwóch takich userów w tasku i potem oboje znikną jak już przyjdą świeże dane z backendu.

Tutaj jako argument do `mutateTask` podajesz zmodyfikowane dane, które mają zostać użyte do wyświetlenia zmian na żywo zanim odświeżą się dane z backendu.

Poniższy kod, znaczy:
Skonstruuj nowy obiekt zawierający wszystko to co `task` (kopię) i nadpisz `task.users`  nową tablicą zawierającą wszystko to co `task.users` (a jak jest puste to stwórz pustą tablicę) z dodanym `user`
`{ ...task, users: [...task?.users || [], user] }`

Tyle, że w tym przypadku ten `user` już występuje w tej liście, jako, że go odpinasz. Powinieneś zatem napisać kod, który konstruuje obiekt, który zawiera userów z usuniętym tym który jest odpinany.

Tutaj GIF na dowód. Ręcznie zatrzymuje serwerem backendowy, żeby zasymulować długi czas odpowiedzi.
![Kapture 2022-06-02 at 22 56 59](https://user-images.githubusercontent.com/31414818/171737160-3ae1a17b-1e21-4258-96b7-df8a5f47be67.gif)

Reply:
Żeby objaśnić ten one-liner.

`task?.users` to jest skrót od:
```js
if(task) task.users
```

To jest tak zwany safe-navigation operator lub optional chaining. Czyli wywoła `task.users` tylko jeśli `task` jest różne od `null` lub `undefined`.

Trzy kropki `...` natomiast służą do 'skopiowania' wszystkich wartości jakiejś tablicy lub obiektu do nowego obiektu/tablicy.

```js
[...task?.users || [], user]
// to samo
[...(task?.users || []), user]
```
`task?.users` zwróci `undefined` jeśli `task` jest `undefined`/`null` lub `task.users` jest `undefined`/`null`.
Problem jest taki, że `[...undefined]` rzuci błędem `Uncaught TypeError: undefined is not iterable`.
Dlatego jest tam operator OR `||`, który uratuje nas przed tym wyjątkiem wstawiająć pustą tablicę jeśli lewa część ekspresji jest `undefined`. Czyli z `[...undefined]` zrobi się `[...[]]` co po prostu daje `[]` pustą tablicę (ale nie rzuca błędem).